[DS-2462] query.filter.spiderIp is redundant, incomplete, scales poorly

mwoodiupui opened the pull request:
## References
Fixes #5823 

## Description
Remove IP-only, agent-only usage grooming.  Use SpiderDetector to make all decisions based on the full array of detectors.

## Instructions for Reviewers
List of changes in this PR:
* Marking and deletion of spider documents by IP only and by agent only is removed from `SolrLoggerService`.
* New method to scan and mark all usage using all of `SpiderDetector`'s capabilities.
* Removed now-unsupported options from `StatisticsClient`.
* Removed unused configuration property from `solr-statistics.cfg`

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
[DS-2462] query.filter.spiderIp is redundant, incomplete, scales poorly

Remove IP-only, agent-only usage grooming.  Use SpiderDetector to make
all decisions based on the full array of detectors.

tdonohue reviewed (approved):
👍 The code here looks generally good to me.  However, I have not tested it, so it might need some testing before merger.  But, I agree with the ideas here of removing detection by IP in favor of just using the SpiderDetector default settings (filtering by DNS and user-agent)

mwoodiupui committed:
[DS-2462] Add testability and starter IT suite to SolrLoggerServiceImpl.
Move SpiderDetectorService javadocs from implementation to interface.
Usual documentation cleanups and minor reformatting.

mwoodiupui committed:
[DS-2462] Add testability and starter IT suite to SolrLoggerServiceImpl.
Move SpiderDetectorService javadocs from implementation to interface.
Usual documentation cleanups and minor reformatting.

mwoodiupui committed:
[DS-2462] Repair merge damage.  #2825

mwoodiupui committed:
[DS-2462] Repair merge damage.  #2825

mwoodiupui committed:
[DS-2462] Fix null IP address from mock, add defensive code for null.

tdonohue reviewed (changes requested):
@mwoodiupui : I tried to test this out today, but when I deploy it, I'm getting a very strange `ClassNotFoundException`:
```
Caused by: org.springframework.beans.factory.CannotLoadBeanClassException: Cannot find class [org.dspace.statistics.​HttpSolrClientFactory] for bean with name 'Ht
tpSolrClientFactory' defined in file [/dspace/config/spring/api/solr-services.xml]; nested exception is java.lang.ClassNotFoundException: org.dspace.statistics.​H
ttpSolrClientFactory
```

Notice that there is somehow a *newline* character between the "H" and the "ttpSolrClientFactory". I didn't add that, and I'm not sure where it's coming from.

But, the end result is the entire 'server' app crashes & therefore the REST API is unavailable.  I'm not sure if it's something on my end or if it'll be reproducible elsewhere.

mwoodiupui committed:
[DS-2462] Remove garbage characters from class name.

mwoodiupui commented:
Three garbage characters crept into the Spring config, I know not how.  Removed.

mwoodiupui committed:
[DS-2462] Don't close Solr core container until all users have finished.

Until 8.9, closing an EmbeddedSolrServer closes the CoreContainer, trashing
any other EmbeddedSolrServer instance that uses the same CoreContainer.
Solr's own tests subclass EmbeddedSolrServer to avoid this.  Now we do too.

tdonohue reviewed (changes requested):
@mwoodiupui : Tried this again today, and I'm still not having any luck.  It's a new error, but the server's still not starting for me:

```
08-Oct-2021 17:50:42.286 SEVERE [localhost-startStop-1] org.apache.catalina.core.ContainerBase.addChildInternal ContainerBase.addChild: start:
        org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/server]]
...
Caused by: java.lang.RuntimeException: Failure during ServletContext initialisation: Error creating bean with name 'org.dspace.statistics.SolrLoggerUsageEventListener#0': Unsatisfied dependency expressed through method 'setSolrLoggerService' parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'solrLoggerService': Unsatisfied dependency expressed through field 'solrClientFactory'; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.dspace.statistics.SolrClientFactory' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}
...
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'org.dspace.statistics.SolrClientFactory' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Autowired(required=true)}
```

All I did was pull your latest PR code into main & build & deploy:
```
git pull
git checkout -b mwoodiupui-DS-2462 main
git pull https://github.com/mwoodiupui/DSpace.git DS-2462
```

Maybe there's something new on `main` that is conflicting?  Or you forgot to commit an update?

mwoodiupui committed:
[DS-2462] Better document workaround from Solr issues.

mwoodiupui committed:
Repair tests after merge.

mwoodiupui committed:
Repair tests after merge.

mwoodiupui committed:
Clean up superseded bean wiring.

mwoodiupui commented:
A lot of stuff changed since this patch was built, but I think I have it caught up.  I was able to start the server.

mwoodiupui committed:
Make Solr field access more rugged.

mwoodiupui committed:
More bean cleanup.

mwoodiupui committed:
Repair rebase damage.

mwoodiupui committed:
Give test access to necessary field.

Also remove an unused field.

mwoodiupui committed:
Merge branch 'main' into DS-2462

mwoodiupui committed:
Satisfy Checkstyle

tdonohue commented on dspace-api/src/main/java/org/dspace/statistics/SolrLoggerServiceImpl.java:
```diff
@@ -772,83 +768,40 @@ public void process(SolrInputDocument doc) throws IOException, SolrServerExcepti
-                @Override
-                public void process(SolrInputDocument doc) throws IOException, SolrServerException {
+    public void markRobots() {
+        ResultProcessor processor = new ResultProcessor() {
```
In reviewing this PR, I'm wondering why the entire `SpiderDetector.getSpiderIpAddresses()` method (replaced on this line) isn't removed? That method seems to no longer be called except by Tests.

Based on the purpose of this PR, it seems like the PR should **also** be deleting all the code related the (outdated) `iplists.com-*.txt` files in this folder https://github.com/DSpace/DSpace/tree/main/dspace/config/spiders   But, maybe I'm misunderstanding the scope of this PR?

Is there are reason to keep around the IP address code when you are replacing its usages?

mwoodiupui replied:
You're right.  Removed.

tdonohue reviewed (commented):
@mwoodiupui : Thanks for updating this again recently. I gave this a look & test today.  Overall, the new code appears to be working (as best I can tell).  However, I was a bit surprised that we are still reliant on the outdated Spider IPs in the `[src]/dspace/config/spiders/iplists.com-*.txt` files.  It looks like those are not removed, and I see that the `SpiderDetector.isSpider(String ip)` method is still used & loads them (via `SpiderDetectorServiceImpl.loadSpiderIpAddresses()`).

Is it purposeful to leave the IP address approach in place for now?  I had assumed that the solution would be to migrate from using IPs to using DNS / user-agent, as implied in #5823 

In any case, I do want to note this seems like it is a basic fix.  It removes the `query.filter.spiderIp` config.  So, I'm OK with this "as-is", if you'd prefer to keep this PR tightly scoped.  In which case, I can approve & merge it....just let me know.

mwoodiupui commented:
I never intended to remove the ability to identify spiders by IP address.  What I'm removing is the code that tests *only* by IP address.  Note in the description:  "full array of detectors".  The problem I'm addressing is that some of our code was using only IP address and thus missing spiders which are known only by domain or agent name.  All checking for spiders should check all available identities.  And if we find yet another way to identify spiders, it can be added centrally and will thereafter be used by all code that cares about spiders without further patching.

Is there some reason not to implement testing by IP address?

mwoodiupui commented:
If you're pointing out that the IP address files are stale, I agree, but that is another matter.

Also, sites may have local lists that they keep up-to-date, including internal spiders such as link testers and vulnerability scanners.

mwoodiupui commented:
The COUNTER list fetcher patch may lead to adding yet another loader, and to new matching logic.  We really need to make all that detail opaque to dependents of this package.

tdonohue commented:
@mwoodiupui : It sounds like I misunderstood the original #5823 issue ticket description then :)   I had thought the mentions of of the flaws of the IP-based approach in that ticket were implying we wanted to remove IP-based checks.  Sounds like that was a misunderstanding, and now that I re-read things, I can see the flaws it mention may be specific to `query.filter.spiderIp` and not specific to *all* IP-based checks.

So, I'm OK with moving this forward as-is.  I do still believe that my question about `SpiderDetector.getSpiderIpAddresses()` is valid.  This method seems to be specific to `query.filter.spiderIp` and is no longer used (except by tests) as this PR removes its only usage.  Should we remove it, to avoid it being used again in the future?  (Personally, I prefer removing unused code, as it is simply clutter.)  Thoughts on that?

mwoodiupui commented:
Ah, well spotted.  getSpiderIpAddresses() is only used by code that is itself no longer used.  I'm hacking all of that out.  And, nothing outside of the spider detection classes should be using, or even aware of, the spider IPs list.

mwoodiupui committed:
Remove dead code.

tdonohue reviewed (approved):
👍 Thanks @mwoodiupui   This looks good to me now.  I'll go ahead and merge it. 

I suspect we should mention the removal of this configuration in the 8.0 Release Notes.  Would you be willing to add a note to the release notes, perhaps in the "Major Updates and Improvements" section? https://wiki.lyrasis.org/display/DSDOC8x/Release+Notes#ReleaseNotes-MajorUpdatesandImprovements
